### PR TITLE
Validate json parse for TargetGroup Unmarshal

### LINKED ISSUE #3365: Scraping both file_sd_configs and dns_sd_configs scrapes only dns
**What did you do?**

```yaml

- job_name: 'foobar'
  scrape_interval: 30s
  file_sd_configs:
    - files:
      -  /etc/prometheus/config/_p_foobar_s__foobar/_p_foobar_s__foobar.json

- job_name: 'servdisc'
  scrape_interval: 30s
  dns_sd_configs:
    - names:
      - _supporthttp._tcp.barfoo-foobar.tld.

```
**What did you expect to see?**

Scraping on both targets

**What did you see instead? Under which circumstances?**

only scraping on dns_sd_configs

**Environment**

* System information:

	 `Linux 4.9.0-3-amd64 x86_64` / debian stretch

* Prometheus version:

```
$ prometheus --version
prometheus, version 1.8.0 (branch: HEAD, revision: 3569eef8b1bc062bb5df43181b938277818f365b)
  build user:       root@bd4857492255
  build date:       20171006-22:12:46
  go version:       go1.9.1
```

* Prometheus configuration file:

```yaml

scrape_configs:

  - job_name: 'foobar'
    scrape_interval: 30s
    file_sd_configs:
      - files:
        -  /etc/prometheus/config/_p_foobar_s__foobar/_p_foobar_s__foobar.json

  - job_name: 'servdisc'
    scrape_interval: 30s
    dns_sd_configs:
      - names:
        - _supporthttp._tcp.barfoo-foobar.tld.

```

the content of the json is :

```json
[{
    "labels": {
        "env": "prod",
        "group": "foo_bar_service",
        "service_type": "webservice"
    },
    "target": [
        "foo01.tld:7646",
        "foo02.tld:7646",
        "foo03.tld:7646",
        "foo04.tld:7646"
    ]
}]
```
if I completely disable dns_sd_configs it is scraped again

### PR DESCRIPTION
**Pull request by krasi-georgiev**
fixes #3365

### CODE REVIEW

**Comment by brian-brazil** [comment]
Can we use the same checkOverflow pattern we use elsewhere for this?

**Comment by krasi-georgiev** [comment]
Thanks for pointing it out.

https://godoc.org/gopkg.in/yaml.v2#UnmarshalStrict
I had a quick look at the yaml package and it had implemented their own Strict Unmarshal-ing so another option would be to upgrade the yaml package so `checkOverflow` will become obsolete and with go 1.10 can also remove the `findUnknownFields` so no extra checks would be needed.

**Comment by krasi-georgiev** [comment]
just double checked the `checkOverflow ` pattern and unfortunately the 'inline' tag is not supported in the json package so can't use it.

**Comment by krasi-georgiev** [comment]
@brian-brazil PTAL

**Comment by brian-brazil** [comment]
Go 1.10 is only two months away, I'm not sure it's worth putting this in in the meantime give that it's not our usual pattern.

@fabxc @juliusv What do ye think?

**Comment by krasi-georgiev** [comment]
Wouldn't be a problem that if the 1.10 strict json parsing is used prometheus would require golang 1.10 and above ?

I think it is better to have something now and refactor when possible rather than wait for 1.10 and risk to forget about this bug all together.

**Comment by brian-brazil** [comment]
> Wouldn't be a problem that if the 1.10 strict json parsing is used prometheus would require golang 1.10 and above ?

That's not a problem. That's only a concern with client_golang.

**Comment by krasi-georgiev** [comment]
@brian-brazil I referenced it incorrectly so shouldn't be closed from #3856 
Still needs to be reviewed and merged.

**Comment by brian-brazil** [comment]
You have test failures

File: discovery/targetgroup/targetgroup_test.go [hunk 1]
```diff
@@ -1,3 +1,16 @@
+// Copyright 2016 The Prometheus Authors
```
Comment (brian-brazil) [inline discovery/targetgroup/targetgroup_test.go]: 2018

**Comment by krasi-georgiev** [comment]
sh*t added some extra files , will fix in few hours.

**Comment by krasi-georgiev** [comment]
@sdurrheimer can you give me a clue why is this still trying to use the 1.9 go builder?

here is the circle file for this branch
https://github.com/krasi-georgiev/prometheus/blob/validate-json-parse/circle.yml

**Comment by krasi-georgiev** [comment]
@SuperQ maybe you can give some pointers on why is this failing?

**Comment by krasi-georgiev** [comment]
@brian-brazil all green, 
thanks for the help , I got lost a bit.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**